src/helpers.c

```diff
@@ -579,7 +612,44 @@ enum swupd_code swupd_init(enum swupd_init_config config)
 	}
 
 	if ((config & SWUPD_NO_NETWORK) == 0) {
-		if (swupd_curl_init() != 0) {
+
+		bool curl_init = true;
+
+		/* try to init curl with whatever server is set up */
+		if (swupd_curl_init(NULL) != 0) {
+			/* the server is not responding */
+			if (mirror_is_set()) {
+				/* the server is a mirror, check to see if the upstream server is working */
+				swupd_curl_deinit();
+				char *upstream_version_url = NULL;
+				get_upstream_version_url(&upstream_version_url);
+				if (upstream_version_url && swupd_curl_init(upstream_version_url) == 0) {
+					/* the upstream server is up, unset the mirror */
+					warn("The mirror '%s' is not responding\n", version_url);
+					info("Removing mirror configuration\n");
+					free_string(&upstream_version_url);
+					unset_mirror_url();
+				} else {
+					/* the upstream server and mirror are not responding,
+					 * this may indicate a connectivity problem with the
+					 * client, so do not unset the mirror at this time */
+					curl_init = false;
+					free_string(&upstream_version_url);
+				}
+			} else {
+				/* the upstream server is not responding and there is no mirror */
+				curl_init = false;
+			}
+		}
+
+		if (!curl_init) {
```

Comment:
this can move inside the if above because curl_init is never set to false outside of that if